LayoutTests/fast/attachment/resources/attachment-test-utils.js

```diff
@@ -0,0 +1,20 @@
+function takeScreenshotWhenAttachmentsSettled()
+{
+    console.assert(document.documentElement.classList.contains("reftest-wait"));
+
+    window.addEventListener("load", async function() {
+        let attachments = Array.from(document.getElementsByTagName("attachment"));
+        while (true) {
+            console.assert(attachments.length);
+            const allIconsLoaded = attachments.every(a => a.hasThumbnailOrIcon);
+            if (allIconsLoaded)
+                break;
+
+            await new Promise(requestAnimationFrame);
+
+            attachments = Array.from(document.getElementsByTagName("attachment"));
+        }
+
+        document.documentElement.classList.remove("reftest-wait");
+    });
+}
```

Comment:
```suggestion
async function takeScreenshotWhenAttachmentsSettled()
{
    console.assert(document.documentElement.classList.contains("reftest-wait"));
    if (document.readyState != "complete")
        await new Promise(() => addEventListener("load", takeScreenshotWhenAttachmentsSettled, { once: true }));
    let attachments = [...document.getElementsByTagName("attachment")];
    console.assert(attachments.length);
    if (attachments.every(a => a.hasThumbnailOrIcon))
        document.documentElement.classList.remove("reftest-wait");
    else
        requestAnimationFrame(takeScreenshotWhenIconLoaded);
}
```
You can write this recursively if you want.

Reply:
I've changed to your style, thanks.

However I've re-added an iteration count, as I found a few tests that timed out. I'll try to debug, but it seems very difficult to get any logging or to attach-to&debug these layout tests!

Reply:
I haven't found the cause of the timeouts, but now if the iteration count goes too high, a comment is inserted in the web page, which makes it fail (as it would have anyway, but quicker) and points at the issue.